flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+intrinsic may return several types and would have a similar implementation, the
+implementation should be done in `lib/Evaluate/fold.cpp`.
+Add the name of the intrinsics in the if-chain, and plug the implementation there.
+See other implementations for examples.
+
+Front-end folding operates on `Fortran::evaluate::Expr`, and in particular, with the
+`Fortran::evaluate::Constant` variant of expressions. You will need to be
+familiar with them to do anything. Look at the defining header in `include/Evaluate`
+to get started. 
+
+#### Tests
+To add regression tests for front-end folding, `test_folding.sh` can be used as a test driver.
+It checks that every parameter prefixed by `test_` is folded to `.true.`.
+`test/Evaluate/folding06.f90` is a good example for it.
+
+## Adding intrinsics to the runtime and lowering
+
+It is first needed to decide whether some runtime is needed or not. If the intrinsic can
+be  implemented with a simple sequence of FIR operations, it might be worth to implement it inline.
+Otherwise, if it involves more advanced control flow and dynamic allocations, a runtime implementation
+might be sounder.
+
+### Runtime
+- See: `lib/runtime` folder
+- Where to submit patch: llvm repository through Phabricator.
+
+#### Implementation
+A runtime interface must be designed for the intrinsic.
+The runtime interface is simply the declaration of the runtime function for which lowering must insert calls,
+as well as a small description of what the function is doing (e.g. comments above the declaration).
+Before starting to implement anything, it is best to submit this interface for review. Having the interface
+already submitted can also allow to implement the lowering part in parallel of the runtime implementation.
+See `lib/runtime/character.h` for examples of runtime interfaces.
+
+The design must decide if one or more functions should form the runtime interface.
+It may make sense to have simple entry points for easy cases, and a more complex entry points
+with descriptors to cover all the other cases. When in doubt, it is best to start with a single
+entry point that cover all cases, and leave entry point specialization for later, based on a measurable
+performance gain on actual programs.
+
+The implementation is done in C++, but there is a hard requirement in not depending on
+the C++ runtime library. That is because it is not wished that Fortran runtime depends
+upon libstdc++ or other C++ runtime libraries. This means that templates and classes
+are OK, `#include <c[library]>` should be OK, but any other includes should be avoided
+unless proven they do not bring C++ runtime in.
+
+The implementation is done in a related `.cpp` file in the same folder. See
+`lib/runtime/character.cpp` for an example.
+
+#### Tests
+To be defined. Ideally, we would want end-to-end tests from Fortran, but it is
+not yet clear where to put them in llvm. Unit tests can also be defined. See
```

Comment:
Should probably go here https://github.com/llvm/llvm-test-suite but that's speculation.

Reply:
I'll leave the To be defined until we are sure.